network: Add support for macvlan and macvtap driver

amshinde opened the pull request:
Add support for macvlan driver by bridging it with a macvtap or
tap+bridge and moving the mac and ip address from the
macvlan to the bridged interface.

Fixes #162

Signed-off-by: Archana Shinde <archana.m.shinde@intel.com>

jodh-intel commented on virtcontainers/network.go:
```diff
@@ -511,6 +581,16 @@ func (n *NetworkNamespace) UnmarshalJSON(b []byte) error {
+			}
+
+			endpoints = append(endpoints, &endpoint)
+			virtLog.Infof("Macvlan endpoint unmarshalled [%v]", endpoint)
```
I'm not sure what format the endpoint is in but it might be better to write this as:

```
virtLog.WithField("endpoint", endpoint).Info("Macvlan endpoint unmarshalled")
```

jodh-intel replied:
^ This comment hasn't been addressed yet :)

amshinde replied:
fixed.

jodh-intel commented on virtcontainers/network.go:
```diff
@@ -933,29 +1029,41 @@ func bridgeNetworkPair(netPair *NetworkInterfacePair) error {
+			return fmt.Errorf("Could not get veth interface %s : %s", netPair.VirtIface.Name, err)
+		}
+	} else {
+		return fmt.Errorf("Unexpected endpointType  %s", endpointType)
```
Nit: two spaces after `endpointType`.

amshinde replied:
fixed.

jodh-intel commented on virtcontainers/network.go:
```diff
@@ -1138,36 +1264,65 @@ func createVirtualNetworkEndpoint(idx int, ifName string, interworkingModel NetI
+	netPair := createNetworkInterfacePair(idx, ifName, interworkingModel)
 
+	endpoint := &BridgedMacvlanEndpoint{
+		// TODO This is too specific. We may need to create multiple
```
Could you either fix this or add an issue URL so we don't forget it.

jodh-intel commented on virtcontainers/network.go:
```diff
@@ -1138,36 +1264,65 @@ func createVirtualNetworkEndpoint(idx int, ifName string, interworkingModel NetI
+func createNetworkInterfacePair(idx int, ifName string, interworkingModel NetInterworkingModel) NetworkInterfacePair {
+	uniqueID := uuid.Generate().String()
+
+	hardAddr := net.HardwareAddr{0x02, 0x00, 0xCA, 0xFE, byte(idx >> 8), byte(idx)}
```
OOI, why are we using these specific values?

In other news, I now feel strangely in need of a coffee... :smile:

amshinde replied:
@jodh-intel  :)
I have just refactored this code, moved this part of existing code into a new function.
Anyways I will add comments in here to explain.

Also, I was planning to have this code merged after the network hotplug PR lands..

jodh-intel commented:
Any update on this @amshinde?

jodh-intel commented:
Hi @amshinde - are you still planning to work on this? It's getting pretty old now... :)

amshinde commented:
@jodh-intel Yes, I am waiting on https://github.com/kata-containers/runtime/pull/623 to be merged first now :) Lot of that PR is touching upon the same files and I would like like to go in first.

raravena80 commented:
Hi @amshinde any updates on this?  thx!

amshinde commented:
@raravena80 Will rebase and update this tomorrow.

amshinde commented:
@raravena80 @jodh-intel @sboeuf I have updated this PR! PTAL.

jodh-intel commented on virtcontainers/network.go:
```diff
@@ -1277,6 +1543,47 @@ func generateInterfacesAndRoutes(networkNS NetworkNamespace) ([]*grpc.Interface,
+		Name: fmt.Sprintf("br%d_kata", idx),
+		VirtIface: NetworkInterface{
+			Name: fmt.Sprintf("eth%d", idx),
+			//HardAddr: hardAddr.String(),
```
Commented out code?

amshinde replied:
fixed

jodh-intel commented on virtcontainers/network.go:
```diff
@@ -1277,6 +1543,47 @@ func generateInterfacesAndRoutes(networkNS NetworkNamespace) ([]*grpc.Interface,
+	return netPair, nil
+}
+
+func generateRandomPrivateMacAddr() (string, error) {
```
This could do with a unit test, particularly since it's cryptographically sensitive.

amshinde replied:
Yup. Will add more unit tests overall.

amshinde replied:
fixed.

jodh-intel commented on virtcontainers/network.go:
```diff
@@ -443,6 +462,138 @@ func (endpoint *PhysicalEndpoint) HotDetach(h hypervisor, netNsCreated bool, net
 	return fmt.Errorf("PhysicalEndpoint does not support Hot detach")
 }
 
+// Macvlan
```
`network.go` is huge - without this PR, it's the largest non-`_test.go` file. With this PR, it is the largest `*.go` file.

I guess it shouldn't block landing the PR, but I'd strongly recommend we split this file into manageable chunks:

- `network.go`
- `network_physical.go`
- `network_vhost.go`
- *etc*

sboeuf replied:
+1 for this proposal. But I would recommend something slightly different. The `network.go` file declare both the `network` interface and the `endpoint` interface.
We should have:
- `network.go`, declaring `network` interface and any generic function related to network.
- `default_network.go` as an implementation of the `network` interface. (here already)
- `noop_network.go` as another implementation of the `network` interface. (here already)
- `endpoint.go` declaring `endpoint` interface and generic functions related to endpoints.
- `macvlan_endpoint.go` defining the MACVLAN implementation.
- `vfio_endpoint.go` defining the VFIO implementation.
and so on...

jodh-intel replied:
Thanks @sboeuf - that indeed looks better! :)

amshinde replied:
@jodh-intel @sboeuf Yup, I realized we need to refactor a lot while working on this PR. I intend to do all of that in a follow up PR. That includes moving each endpoint in its own go file. I did not want to make this PR any larger. I am going to create an issue for the refactor.

amshinde replied:
@jodh-intel @sboeuf https://github.com/kata-containers/runtime/issues/799

jodh-intel replied:
Thanks @amshinde! :)

jodh-intel commented on virtcontainers/network.go:
```diff
@@ -443,6 +462,138 @@ func (endpoint *PhysicalEndpoint) HotDetach(h hypervisor, netNsCreated bool, net
+// tap interface of the network pair to the hypervisor.
+func (endpoint *BridgedMacvlanEndpoint) Attach(h hypervisor) error {
+	networkLogger().Info("Attaching macvlan endpoint")
+	//if err := xconnectVMNetwork(&(endpoint.NetPair), true, endpoint.EndpointType); err != nil {
```
Commented out code.

amshinde replied:
fixed

jodh-intel commented on virtcontainers/network.go:
```diff
@@ -443,6 +462,138 @@ func (endpoint *PhysicalEndpoint) HotDetach(h hypervisor, netNsCreated bool, net
+// Detach for physical endpoint unbinds the physical network interface from vfio-pci
+// and binds it back to the saved host driver.
+func (endpoint *MacvtapEndpoint) Detach(netNsCreated bool, netNsPath string) error {
+	// Bind back the physical network interface to host.
```
This comment suggests the function is actually going to do something... but it doesn't :) Could you make that clearer please?

jodh-intel commented on virtcontainers/network.go:
```diff
@@ -1160,29 +1386,69 @@ func createVirtualNetworkEndpoint(idx int, ifName string, interworkingModel NetI
+	// Another option is to defer this to ciao qemu library which does have
+	// global context but cannot handle errors when setting up the network
+	var err error
+	//endpoint.VMFds, err = createMacvtapFds(endpoint.EndpointProperties.Iface.Index, defaultQueues)
```
Commented out code.

jodh-intel commented on virtcontainers/network.go:
```diff
@@ -1160,29 +1386,69 @@ func createVirtualNetworkEndpoint(idx int, ifName string, interworkingModel NetI
+	// We do not have global context here, hence a manifest constant
+	// that matches our minimum vCPU configuration
+	// Another option is to defer this to ciao qemu library which does have
+	// global context but cannot handle errors when setting up the network
```
Sorry, but I have not idea what this comment is talking about...? :)

jodh-intel commented on virtcontainers/network.go:
```diff
@@ -1160,29 +1386,69 @@ func createVirtualNetworkEndpoint(idx int, ifName string, interworkingModel NetI
+		return nil, fmt.Errorf("Could not setup macvtap fds %s: %s", endpoint.EndpointProperties.Iface.Name, err)
+	}
+
+	//vhostFds, err := createVhostFds(defaultQueues)
```
Another comment to zap.

sboeuf commented on virtcontainers/qemu_arch_base.go:
```diff
@@ -458,6 +458,23 @@ func (q *qemuArchBase) appendNetwork(devices []govmmQemu.Device, endpoint Endpoi
+				MACAddress:    ep.NetPair.TAPIface.HardAddr,
+				DownScript:    "no",
+				Script:        "no",
+				VHost:         true,
```
Does `vhost` have to be true for this case?

Also, except vhost, those two structures look identical to me, IMO you could simply do:
```go
switch ep := endpoint.(type) {
 	case *VirtualEndpoint:
                fallthrough
        case *BridgedMacvlanEndpoint:
                ...
```

sboeuf replied:
Sorry, I think `fallthrough` is not appropriate here, but I think you can do:
```go
switch ep := endpoint.(type) {
 	case *VirtualEndpoint, *BridgedMacvlanEndpoint:
                ...
```

amshinde replied:
Has been addressed.

sboeuf commented on virtcontainers/network.go:
```diff
@@ -1277,6 +1432,28 @@ func generateInterfacesAndRoutes(networkNS NetworkNamespace) ([]*grpc.Interface,
 	return ifaces, routes, nil
 }
 
+func createNetworkInterfacePair(idx int, ifName string, interworkingModel NetInterworkingModel) NetworkInterfacePair {
```
Is the index in the codebase correctly updated? I am asking cause otherwise the hardware address will end up being not unique.

amshinde replied:
Yes. It is passed correctly.

sboeuf commented on virtcontainers/qemu_arch_base.go:
```diff
@@ -475,6 +475,24 @@ func (q *qemuArchBase) appendNetwork(devices []govmmQemu.Device, endpoint Endpoi
+				MACAddress:    ep.HardwareAddr(),
+				DownScript:    "no",
+				Script:        "no",
+				VHost:         true,
```
Same comment as before, couldn't we use `fallthrough` here instead?
I can see the difference is `Type: govmmQemu.MACVTAP,` in this case, but we need a simple helper function to keep this generic. The same way virtual endpoint and macvlan uses `networkModelToQemuType()`.

amshinde replied:
We agreed offline that this cannot be generalized further.

amshinde commented:
/retest

amshinde commented:
@sboeuf @jodh-intel I have addressed your comments. Can you take a look.
Meanwhile I am going to see if I can add some more testing for the networking code in general.

amshinde commented:
/retest

jodh-intel commented on virtcontainers/network.go:
```diff
@@ -443,6 +477,144 @@ func (endpoint *PhysicalEndpoint) HotDetach(h hypervisor, netNsCreated bool, net
+// Attach for virtual endpoint bridges the network pair and adds the
+// tap interface of the network pair to the hypervisor.
+func (endpoint *BridgedMacvlanEndpoint) Attach(h hypervisor) error {
+	networkLogger().Info("Attaching macvlan endpoint")
```
@amshinde - when you refactor the network code, I think it would be great if you could move these useful "attach" / "detach" log messages to the caller so we can make these functions smaller and we won't then have to worry about the consistency of the log messages for the different network drivers.

amshinde replied:
Sure. Will do.

amshinde commented:
/retest

jodh-intel commented on virtcontainers/network_test.go:
```diff
@@ -599,3 +688,106 @@ func TestGenerateInterfacesAndRoutes(t *testing.T) {
+func TestGenerateRandomPrivateMacAdd(t *testing.T) {
+	assert := assert.New(t)
+
+	addr, err := generateRandomPrivateMacAddr()
```
Thanks for adding. Could you also make a second call to `generateRandomPrivateMacAddr()` so that we can check to ensure the results are different?

amshinde replied:
@jodh-intel Done. I have addressed all of @sboeuf's comments. For those pending we had a discussion about it offline. I have added comments for reference.

amshinde commented on virtcontainers/network.go:
```diff
@@ -194,6 +194,15 @@ type VhostUserEndpoint struct {
+type BridgedMacvlanEndpoint struct {
+	NetPair            NetworkInterfacePair
+	EndpointProperties NetworkInfo
+	Physical           bool
```
Has been addressed.

amshinde commented on virtcontainers/network.go:
```diff
@@ -740,23 +843,23 @@ func getLinkByName(netHandle *netlink.Handle, name string, expectedLink netlink.
 
 // The endpoint type should dictate how the connection needs to be made
-func xconnectVMNetwork(netPair *NetworkInterfacePair, connect bool, numCPUs uint32, disableVhostNet bool) error {
+func xconnectVMNetwork(netPair *NetworkInterfacePair, connect bool, numCPUs uint32, disableVhostNet bool, endpointType EndpointType) error {
```
Has been addressed.

amshinde commented on virtcontainers/network.go:
```diff
@@ -843,27 +946,39 @@ func setIPs(link netlink.Link, addrs []netlink.Addr) error {
-	if err != nil {
-		return fmt.Errorf("Could not get veth interface: %s: %s", netPair.VirtIface.Name, err)
+	var link netlink.Link
+
```
Has been addressed.

amshinde commented on virtcontainers/network.go:
```diff
@@ -955,29 +1070,41 @@ func bridgeNetworkPair(netPair *NetworkInterfacePair, numCPUs uint32, disableVho
-		return fmt.Errorf("Could not get veth interface %s : %s", netPair.VirtIface.Name, err)
+	var attrs *netlink.LinkAttrs
+	var link netlink.Link
+
```
Has been addressed.

amshinde commented on virtcontainers/network.go:
```diff
@@ -1029,22 +1156,33 @@ func untapNetworkPair(netPair NetworkInterfacePair) error {
-	} else {
-		if err := netHandle.LinkSetDown(vethLink); err != nil {
-			return fmt.Errorf("Could not disable veth %s: %s", netPair.VirtIface.Name, err)
+	var link netlink.Link
```
Has been addressed.

amshinde commented on virtcontainers/network.go:
```diff
@@ -469,6 +553,9 @@ func (endpointType *EndpointType) Set(value string) error {
 	case "vhost-user":
 		*endpointType = VhostUserEndpointType
 		return nil
+	case "macvlan":
```
macvlan is more appropriate it since it refers to the underlying interface.

jodh-intel commented:
Thanks @amshinde!

lgtm

amshinde commented:
/retest
